### PR TITLE
Attempt to fix FFMPEG 5.0 compatibility

### PR DESCRIPTION
**Pull request by bjuncek**
Attempting to address #5616 via simple API update.
Should be properly tested in FB internals.  

cc @datumbox

### CODE REVIEW

File: torchvision/csrc/io/decoder/subtitle_stream.cpp
Comment (datumbox) [inline torchvision/csrc/io/decoder/subtitle_stream.cpp:53]: Nit: This code is repeated on the decoder.cpp file. Would it make sense to move the initialisation to a separate internal utility method?

**Review by datumbox — COMMENTED** [review]
@bjuncek Thanks for fixing this. The changes look good to me. I've only added one comment for your attention.

Our CI is currently broken, so I propose not to merge this until #5648 is merged and it's properly tested on FB infra as you suggested. We'll get back to you on the latter.

File: torchvision/csrc/io/decoder/decoder.cpp
Comment (jdsgomes) [inline torchvision/csrc/io/decoder/decoder.cpp:489]: according to LeakSanitizer this call will result in a memory leak.

**Review by jdsgomes — COMMENTED** [review]
I have imported the PR and pointed out a few things to be considered.

File: torchvision/csrc/io/decoder/decoder.cpp
Comment (datumbox) [reply torchvision/csrc/io/decoder/decoder.cpp:489]: We should call `av_packet_free(&avPacket);` when we are done with it. We need to be careful and add this to all the places where we return early too.

**Comment by jdsgomes** [comment]
@bjuncek I made a few changes in an attempt to fix the memory leak and the static code analysis warnings, but there are a few things that came up that would need discussion when you are back as you will probably have more context:

- if you agree with the approach [here](t[orchvision/csrc/io/decoder/subtitle_stream.cpp](https://github.com/pytorch/vision/pull/5644/files#diff-7733b1c80a95250516032ddd798d930100e5348ecdb6db558cc789acf11ad773R591)) and [[torchvision/csrc/io/decoder/subtitle_stream.cpp](https://github.com/pytorch/vision/pull/5644/files#diff-7733b1c80a95250516032ddd798d930100e5348ecdb6db558cc789acf11ad773R493)](url), we probably need to follow the same in `torchvision/csrc/io/decoder/subtitle_stream.cpp`
- it is not 100% clear to me what is the difference between `av_packet_unref` and `av_packet_free`. From what I can gather  in the documentation, the former unreferences the inner buffer and resets the default values and the later frees the actual avPacket object but please confirm that is the case and that we need both in the way it has been used

File: torchvision/csrc/io/decoder/decoder.cpp
Comment (bjuncek) [reply torchvision/csrc/io/decoder/decoder.cpp:490]: Yeah, I was thinking about this, but my thought was that `av_packet_unref` would have freed whatever was allocated (which seems to have been an incorrect assumption). 

I'll fix this ASAP

**Comment by bjuncek** [comment]
>> if you agree with the approach [here](https://github.com/pytorch/vision/pull/5644/files#diff-7733b1c80a95250516032ddd798d930100e5348ecdb6db558cc789acf11ad773R493) and [here](https://github.com/pytorch/vision/pull/5644/files#diff-7733b1c80a95250516032ddd798d930100e5348ecdb6db558cc789acf11ad773R502), we probably need to follow the same in torchvision/csrc/io/decoder/subtitle_stream.cpp

Yup, have added the changes; this is something you might need checking in FB infra as subtitle stream was (yet) not used in TV, but might have been used internally (I've been long gone). 

>> it is not 100% clear to me what is the difference between av_packet_unref and av_packet_free. From what I can gather in the documentation, the former unreferences the inner buffer and resets the default values and the later frees the actual avPacket object but please confirm that is the case and that we need both in the way it has been used

My understanding in this case is the same as yours, and I do thing we'd need both; do these changes pass the static memory leak checks?

File: torchvision/csrc/io/decoder/decoder.cpp
Comment (datumbox) [inline torchvision/csrc/io/decoder/decoder.cpp:578]: Two points on this:
1. This second unref was originally here because the loop might terminate via a break. The code is messy and we should refactor it to make it easier to read and have better flow. That's a future NIT.
2. Given that now we call `av_packet_free` call few lines below, should we drop the unref call?

File: torchvision/csrc/io/decoder/subtitle_stream.cpp
Comment (datumbox) [inline torchvision/csrc/io/decoder/subtitle_stream.cpp:50]: See Joao's comment on the `nullptr`. Also, should we be returning an error if allocating a packet fails? Finally it's unclear to me why we allocate a packet even when `packet` is not null. This can be improved on a future PR.

File: torchvision/csrc/io/decoder/decoder.cpp
Comment (bjuncek) [reply torchvision/csrc/io/decoder/decoder.cpp:578]: 1. I agree. IMO the entire API needs a refactor, ideally to be merged with the GPU decoding; This approach  is overly complicated (for historical reasons) for our purposes.
2. Yes if the documentation is to be believed; there are two versions of free packet: `av_packet_free` (which does un-referencing for us), and `av_free_packet` (that doesn't)

File: torchvision/csrc/io/decoder/subtitle_stream.cpp
Comment (bjuncek) [reply torchvision/csrc/io/decoder/subtitle_stream.cpp]: small enough change that I can push it now I take it, and revert if it breaks something profoundly

File: torchvision/csrc/io/decoder/subtitle_stream.cpp
Comment (datumbox) [inline torchvision/csrc/io/decoder/subtitle_stream.cpp:55]: Are we sure that this error code will work as return type for the method? Should we be returning one of the error codes of avcodec instead?

File: torchvision/csrc/io/decoder/subtitle_stream.cpp
Comment (datumbox) [inline torchvision/csrc/io/decoder/subtitle_stream.cpp:71]: Why is this variable in scope?

File: torchvision/csrc/io/decoder/subtitle_stream.cpp
Comment (bjuncek) [reply torchvision/csrc/io/decoder/subtitle_stream.cpp]: I've updated them to the most appropriate [AVERROR](https://ffmpeg.org/doxygen/trunk/group__lavu__error.html)


Having done that, I'd say ENOMEM is more appropriate than AV alternative that I see used (`AVERROR_BUFFER_TOO_SMALL`).

File: torchvision/csrc/io/decoder/subtitle_stream.cpp
Comment (datumbox) [inline torchvision/csrc/io/decoder/subtitle_stream.cpp:85]: We need to clean up `av_packet_free(&avPacket);` prior returning here.

File: torchvision/csrc/io/decoder/subtitle_stream.cpp
Comment (datumbox) [inline torchvision/csrc/io/decoder/subtitle_stream.cpp:54]: This is close enough for error code I think.

File: torchvision/csrc/io/decoder/decoder.cpp
Comment (datumbox) [inline torchvision/csrc/io/decoder/decoder.cpp:493]: In this method, this is not a supported error code (at least according to its doc). We should either update the doc to document this error code or adopt a more fitting enum for this.

File: torchvision/csrc/io/decoder/decoder.cpp
Comment (datumbox) [inline torchvision/csrc/io/decoder/decoder.cpp:502]: Are you sure we should be freeing the memory here? I think this callable is used in the loop below, it doesn't actually return for the whole method.

File: torchvision/csrc/io/decoder/decoder.cpp
Comment (bjuncek) [reply torchvision/csrc/io/decoder/decoder.cpp]: is this better?

**Review by datumbox — COMMENTED** [review]
Thanks for your work @bjuncek. I think it looks good.

@jdsgomes Could you check if this works OK internally?

File: torchvision/csrc/io/decoder/subtitle_stream.cpp
Comment (jdsgomes) [inline torchvision/csrc/io/decoder/subtitle_stream.cpp:47]: agree - I am also not sure if this should be created. Is the plan to follow up on this?

File: torchvision/csrc/io/decoder/subtitle_stream.cpp
Comment (datumbox) [reply torchvision/csrc/io/decoder/subtitle_stream.cpp:54]: @jdsgomes The two methods return different enums. Originally Bruno had them synced; no strong opinions on my side.

File: torchvision/csrc/io/decoder/subtitle_stream.cpp
Comment (bjuncek) [reply torchvision/csrc/io/decoder/subtitle_stream.cpp:47]: Yes, but I think this would require a bit of coordination on Meta's side, as this is not used nor tested in TV.

File: torchvision/csrc/io/decoder/subtitle_stream.cpp
Comment (bjuncek) [reply torchvision/csrc/io/decoder/subtitle_stream.cpp:54]: @jdsgomes would you like me to change it to match?

Comment (jdsgomes) [reply torchvision/csrc/io/decoder/subtitle_stream.cpp:54]: I would prefer for consistency, unless there is a reason not to. Other than that the PR looks good and the internal tests are passing. Thank you for these changes

**Comment by jdsgomes** [comment]
For reference the internal tests are passing: https://www.internalfb.com/diff/D35115411

**Comment by bjuncek** [comment]
@datumbox I'll leave it up to you to decide if it's ready for merge.

**Review by jdsgomes — APPROVED** [review]

**Comment by jdsgomes** [comment]
I have a nit regarding he return codes but definitely non blocking.

**Review by datumbox — APPROVED** [review]
@bjuncek LGTM, thanks!

@jdsgomes I agree that a good API should align the return error codes across functions. I'm OK if you want to merge this now and follow up in a new PR. I'm also OK if you want to push the change directly on this branch. I leave it to you how we should proceed. At any case, we should merge this soon. We are already a week with broken CI and the issue is masking other potential problems that are not reported to the release engineering team.

File: torchvision/csrc/io/decoder/decoder.cpp
Comment (jdsgomes) [inline torchvision/csrc/io/decoder/decoder.cpp:493]:
```suggestion
    return AVERROR_BUFFER_TOO_SMALL;
```

File: torchvision/csrc/io/decoder/decoder.cpp
Comment (jdsgomes) [inline torchvision/csrc/io/decoder/decoder.cpp:4]:
```suggestion
```

**Pull request merged**
